prow/tide/contextregister.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tide
+
+import (
+	"sync"
+
+	"github.com/shurcooL/githubql"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type ContextChecker interface {
+	// IgnoreContext tells whether a context is optional.
+	IgnoreContext(context Context) bool
+	// MissingRequiredContexts tells if required contexts are missing from the list of contexts provided.
+	MissingRequiredContexts([]Context) []Context
+}
+
+// newExpectedContext creates a Context with Expected state.
+// This should not only be used when contexts are missing.
+func newExpectedContext(c string) Context {
+	return Context{
+		Context:     githubql.String(c),
+		State:       githubql.StatusStateExpected,
+		Description: githubql.String(""),
+	}
+}
+
+// ContextRegister implements ContextChecker and allow registering of required and optional contexts.
+type ContextRegister struct {
+	lock               sync.RWMutex
```

Comment:
I don't understand the need for thread safety... I don't see where we concurrently register required contexts

Reply:
I don't see why it would hurt anything. Tide runs multiple things in goroutine.